app/src/server/stats.ts

```diff
@@ -0,0 +1,242 @@
+    }
+  )
+}
+
+/**
+ * Initialize classification stats
+ * @param categories list of possible categories
+ * @param attributes list of possible attributes
+ * @return map of category and attribute counts initialized to 0
+ */
+function initClassificationStats (
+  categories: string[], attributes: Attribute[]):
+  ClassificationStats {
+  const categoriesByName = _.keyBy(categories)
+  return _.mapValues(categoriesByName,
+    (_category) => {
+      return {
+        count: 0,
+        attributeCounts: initAttributeCount(attributes)
+      }
+    })
+}
+
+/**
+ * Get the value of an attribute given the index in the value list
+ */
+function getAttributeValue (attribute: Attribute, index: number) {
+  if (attribute.toolType === AttributeToolType.SWITCH) {
+    return index === 1 ? 'true' : 'false'
+  }
+  return attribute.values[index]
+}
+
+/**
+ * Update the classification stats with the given label
+ * @param stats the stats so far
+ * @param label the new label
+ * @param categories the list of possible categories
+ * @param attributes the list of possible attributes
+ */
+function updateClassificationStats (
+  stats: ClassificationStats, label: LabelType,
+  categories: string[], attributes: Attribute[]): ClassificationStats {
+  const result = stats
+  for (const categoryIndex of label.category) {
+    const categoryName = categories[categoryIndex]
+    result[categoryName].count += 1
+    for (const attributeKey of Object.keys(label.attributes)) {
+      const attributeIndex = Number(attributeKey)
+      const attr = attributes[attributeIndex]
+
+      for (const attributeValueIndex of label.attributes[attributeIndex]) {
+        const value = getAttributeValue(attr, attributeValueIndex)
+        result[categoryName].attributeCounts[attr.name][value] += 1
+      }
+    }
+  }
+  return result
+}
+
+/**
+ * Get the label breakdown by class
+ */
+export function getClassificationStats (
+  tasks: TaskType[]): ClassificationStats {
+  if (tasks.length === 0) {
+    return {}
+  }
+
+  const config = tasks[0].config
+  const categories = config.categories
+  const attributes = config.attributes
+
+  let result = initClassificationStats(categories, attributes)
+
+  for (const task of tasks) {
+    for (const item of task.items) {
+      for (const label of Object.values(item.labels)) {
+        result = updateClassificationStats(
+          result, label, categories, attributes)
+      }
+    }
+  }
+  return result
+}
+
+/** the number of labels for each attribute type/value */
+interface AttributeCounts {
+  [name: string]: { [value: string]: number}
+}
+
+/**
+ * Stats for a single category
+ */
+interface CategoryStats {
+  /** the number of labels with the category */
+  count: number
+  /** the counts for each attribute within the category */
+  attributeCounts: AttributeCounts
+}
+
+/**
+ * Stats for all classification categories
+ */
+export interface ClassificationStats {
+  [name: string]: CategoryStats
+}
```

Comment:
1. We can simplify the names and class relations here. It is better to mirror the label state structure in the stats. 

2. The meaning of `SomethingStats` is intuitively statistics of something. The code will also be easier to read. 
3. `CategoryStats` can manage the statistics of the categories instead of one category, so that `LabelStats` doesn't have to concern category names.

```suggestion
export interface LabelStats {
  category: CategoryStats,
  attribute: AttributeStats
}
```

Reply:
I took your suggestion and changed the structure to mirror the label state structure. I think the code is clearer now.